### PR TITLE
API Convert installer to recipe

### PR DESCRIPTION
**Pull request by tractorcow**
Implements https://github.com/silverstripe/silverstripe-framework/issues/6860

Recipe is composed as follows:

 -  [recipe-plugin](https://github.com/silverstripe/recipe-plugin): Provides recipe project file installation and additional composer functionality for managing recipes.
 - [recipe-core](https://github.com/silverstripe/recipe-core): Basic mysite/_config.php, .htaccess, web.config. Register project as mysite
 - [recipe-cms](https://github.com/silverstripe/recipe-cms): mysite/code/Page.php and mysite/code/PageController.php
 - `installer` (this) Include mostly-optional and less critical files. Also includes a default theme, and a theme.yml to register it, as well as install and project bootstrap files in root.
 
 Note: After this change, it'll be possible to do `composer require silverstripe/installer 4.0.x-dev` into any project and it'll become a silverstripe application. :D

### CODE REVIEW

**Comment by tractorcow** [comment]
Install log:

image

image

**Comment by tractorcow** [comment]
Hello, could @silverstripe/core-team please give their feedback on this?

**Comment by tractorcow** [comment]
Important discussion on https://github.com/silverstripe/silverstripe-installer/pull/193 that should block release of this feature.

We are thinking of renaming 'project-files' to 'recipe-files', and adding a 'recipe-modules' section.

**Comment by tractorcow** [comment]
Proposed new composer.json schema:

 ```json
 {
    "name": "silverstripe/installer",
    "type": "silverstripe-recipe",
    "description": "The SilverStripe Framework Installer",
    "require": {
        "php": ">=5.6.0",
        "silverstripe/recipe-plugin": "^0.1",
        "silverstripe/recipe-cms": "1.0.x-dev",
        "silverstripe-themes/simple": "~3.2.0"
    },
    "require-dev": {
        "phpunit/PHPUnit": "^5.7"
    },
    "extra": {
    	"project-dependencies-installed": [
    		"silverstripe/cms": "4.0.x-dev",
    		"silverstripe/framework": "4.0.x-dev"
    	],
    	"project-files-installed": [
    		"index.php",
    		"install.php"
    	],
        "project-files": [
            "mysite/_config/*",
            "index.php",
            "install.php",
            ".env.example",
            "favicon.ico",
            "install-frameworkmissing.html"
        ],
        "branch-alias": {
            "4.x-dev": "4.0.x-dev",
            "dev-master": "5.x-dev"
        }
    }
}
 ```
 
 Notes:
 
- Project-files is only used by recipes being included. This element is actually deleted in the root project after install, since it's no longer necessary if this is the root project (you can't upgrade a root project anyway). See https://github.com/silverstripe/recipe-plugin/commit/0668b1d0fc245f8ab9a0636161ac288067b96019
 - project-files-installed is populated from all project files installed from NESTED recipes. E.g. if installer recipe requires recipe-core, any file recipe-core installs will be registered here. If this file is manually removed, the recipe plugin will see the existing reference and avoid re-installing it from future updates to recipe-core.
 - project-dependencies-installed will work the same way, but ONLY when inlining a recipe. E.g. if you never inline recipe-core, you won't see project-dependencies-installed populated. This is because project files works whether or not you inline a dependency, but nested dependencies are only inlined if you call require-recipe / update-recipe.

**Comment by tractorcow** [comment]
Ok, I've implemented https://github.com/silverstripe/recipe-plugin/pull/3 which should address the removablity of modules and files. Once this is merged it'll be possible to inline a recipe, remove a dependency, and update it to newer versions without the removed items being re-installed.

File: composer.json [hunk 1]
```diff
@@ -1,26 +1,25 @@
...
         "phpunit/PHPUnit": "^5.7"
     },
     "extra": {
+        "project-files": [
+            "mysite/_config/*",
+            "index.php",
+            "install.php",
+            ".env.example",
+            "favicon.ico",
+            "install-frameworkmissing.html"
```
Comment (flamerohr) [inline composer.json]: what about the `assets` folder and files? Might be important to included the `web.confg` and `.htaccess` files at least.
Or perhaps they belong in `recipe-core`, which seems more logical.

Comment (tractorcow) [inline composer.json]: `web.config` and `.htaccess` are provided by nested recipes; A parent recipe shouldn't re-declare files from children as it's not necessary.

I think you are right about assets folder though; I'll add that to the list.

Comment (tractorcow) [inline composer.json]: Added.

**Comment by flamerohr** [comment]
the Behat test failing in asset-admin is a pre-existing broken test, will merge on the condition that it will be fixed soon-ish

**Pull request merged**